lib/nmatrix/shortcuts.rb

```diff
@@ -34,6 +34,92 @@
 
 class NMatrix
 
+  # Methods for generating magic matrix.
+  module MagicHelpers
+    class << self
+      def odd_magic(nm,shape)
+        row= shape-1
+        col= shape/2       
+        nm[row,col]= 1
+        (2..shape*shape).each do |i|
+          if nm[(row+1)%shape,(col+1)%shape] == 0
+            row= (row+1)%shape
+            col= (col+1)%shape
+          else
+            row= (row-1+shape)%shape
+          end
+            nm[row,col]= i
+        end
+      end
+    
+      def doubly_even_magic(nm,shape)
+        mini_square_num = shape/4
+        cnt = 1     
+        inv_cnt = shape*shape
+        shape.times do |i|
+          shape.times do |j|
+            if j>=mini_square_num && j<shape-mini_square_num
+		      if i>=mini_square_num && i<shape-mini_square_num
+      	        nm[i,j] = cnt
+		      else 
+		        nm[i,j] = inv_cnt
+		      end
+		    elsif i<mini_square_num || i>=shape-mini_square_num
+		      nm[i,j]=cnt
+		    else
+              nm[i,j] = inv_cnt
+            end
+          cnt+=1
+          inv_cnt-=1  
+          end
+        end
+      end
+    
+      def singly_even_magic(nm,shape)
+        half_shape = shape/2
+	    k = (shape-2)/4
+	    swap_col= NMatrix.new([shape], dtype: :int32)
+        index=0 
+        mini_magic= NMatrix.new([half_shape,half_shape],0)
```

Comment:
Yes, here the algorithm depends on the matrix being initialized to 0. Absence of which will give incorrect answers.